designs/generic-git-provider-gitops.md

```diff
@@ -0,0 +1,205 @@
+
+### `Git` GitOpsConfig  Spec
+I propose modifying the `GitOpsConfig`, to move the `fluxSystemNamespace` and `clusterConfigPath` fields up a level to be a part of `flux`, instead of reproducing them in both `git` and `github`. provider blocks. This will make for a more consistent experience across provider, all of which will require a `fluxSystemNamespace` and `clusterConfigPath`.
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: GitOpsConfig
+metadata:
+  name: my-generic-git-provider-gitops
+spec:
+  flux:
+    fluxSystemNamespace: ""
+    clusterConfigPath: ""
+    git:
+      repositoryUrl: myClusterGitopsRepo
+      username: myGitProviderUserName
+      passwordEnvironmentVariable: MY_PASSWORD_ENV_VAR
+      privateKeyFile: '/path/to/my/private_key.rsa'
+      branch: branchToUseInRemoteRepo
+```
+
+### Git Configuration Structs
+The `Flux` configuration will be modified to contain generic flux options `FluxSystemNamespace`, `Branch`, and `ClusterConfigPath`, which will be universal across git providers used in conjunction with Flux. A new field, `Git`, will be added, and validations will ensure that `Git` and `Github` (and any other Provider type) are mutually exclusive.
+```Go
+type Flux struct {
+    // Github is the name of the Git Provider to host the Git repo.
+	Github Github `json:"github,omitempty"`
+
+    // Git is the name of the generic Git Provider to host the Git repository
+    Git Git `json:"git,omitempty"`
+
+    // FluxSystemNamespace scope for this operation. Defaults to flux-system.
+    FluxSystemNamespace string `json:"fluxSystemNamespace,omitempty"`
+    
+    // ClusterConfigPath relative to the repository root, when specified the cluster sync will be scoped to this path.
+    ClusterConfigPath string `json:"clusterConfigPath,omitempty"`
+
+	// Git branch. Defaults to main.
+	// +kubebuilder:default:="main"
+	Branch string `json:"branch,omitempty"`
+}
+```
+
+The `Git` struct contains the requisite fields for authenticating to a generic git repository using basic auth, either password, SSH or both together. Use of `net/url` for the repository URL provides out-of-the-box validations and assurance of structure.
+```Go
+type Git struct {
+    // Username is the user to authenticate to the git repository with.
+    Username string `json:username`
+
+    // Repository URL for the repository to be used with flux. Can be either an SSH or HTTPS url.
+    RepositoryUrl net/url `json:repositoryUrl`
+
+    // Path to a private key file for authenticating to the repository via SSH.
+    PrivateKeyFile string `json:privateKeyFile,omitempty`
```

Comment:
how will this file be made available on the workload cluster?

Reply:
updating doc to address this